pandas/core/categorical.py

```diff
@@ -2066,3 +2066,48 @@ def _factorize_from_iterables(iterables):
         # For consistency, it should return a list of 2 lists.
         return [[], []]
     return map(list, lzip(*[_factorize_from_iterable(it) for it in iterables]))
+
+
+class CategoricalType(CategoricalDtype):
+    """
```

Comment:
Is that OK with the caching and stuff that is done on `CategoricalDType`? This is on `CategoricalyDtype`:

```python
    def __new__(cls):

        try:
            return cls._cache[cls.name]
        except KeyError:
            c = object.__new__(cls)
            cls._cache[cls.name] = c
            return c
```

I haven't messed with extension types much. We could make the keys of that internal dict reflect the categories and ordered attributes.